File: frontend/src/lib/api/sns.ts
import { apiClient } from "./client";
import type { Post, PostCreateInput, Comment, CommentCreateInput, PostFilters, Follow } from "@/types/sns";
import type { PaginatedResponse } from "@/types/common";

// Posts
export async function getPosts(
  filters?: PostFilters,
  page = 1,
  limit = 20
): Promise<PaginatedResponse<Post>> {
  const { data } = await apiClient.get("/sns/posts", {
    params: { ...filters, page, limit },
  });
  return data;
}

export async function getPost(id: string): Promise<Post> {
  const { data } = await apiClient.get(`/sns/posts/${id}`);
  return data;
}

export async function createPost(input: PostCreateInput): Promise<Post> {
  const { data } = await apiClient.post("/sns/posts", input);
  return data;
}

export async function deletePost(id: string): Promise<void> {
  await apiClient.delete(`/sns/posts/${id}`);
}

export async function likePost(id: string): Promise<void> {
  await apiClient.post(`/sns/posts/${id}/like`);
}

export async function unlikePost(id: string): Promise<void> {
  await apiClient.delete(`/sns/posts/${id}/like`);
}

// Comments
export async function getComments(
  postId: string,
  page = 1,
  limit = 20
): Promise<PaginatedResponse<Comment>> {
  const { data } = await apiClient.get(`/sns/posts/${postId}/comments`, {
    params: { page, limit },
  });
  return data;
}

export async function createComment(
  postId: string,
  input: CommentCreateInput
): Promise<Comment> {
  const { data } = await apiClient.post(`/sns/posts/${postId}/comments`, input);
  return data;
}

export async function deleteComment(postId: string, commentId: string): Promise<void> {
  await apiClient.delete(`/sns/posts/${postId}/comments/${commentId}`);
}

// Follows
export async function followUser(userId: string): Promise<Follow> {
  const { data } = await apiClient.post(`/sns/users/${userId}/follow`);
  return data;
}

export async function unfollowUser(userId: string): Promise<void> {
  await apiClient.delete(`/sns/users/${userId}/follow`);
}

export async function getFollowers(
  userId: string,
  page = 1,
  limit = 20
): Promise<PaginatedResponse<Follow>> {
  const { data } = await apiClient.get(`/sns/users/${userId}/followers`, {
    params: { page, limit },
  });
  return data;
}

export async function getFollowing(
  userId: string,
  page = 1,
  limit = 20
): Promise<PaginatedResponse<Follow>> {
  const { data } = await apiClient.get(`/sns/users/${userId}/following`, {
    params: { page, limit },
  });
  return data;
}

File: backend/app/api/v1/endpoints/sns.py
from fastapi import APIRouter, Depends, HTTPException, Query
from typing import Optional
from app.schemas.sns import (
    PostResponse,
    PostCreate,
    PostListResponse,
    CommentResponse,
    CommentCreate,
    CommentListResponse,
    FollowResponse,
    FollowListResponse,
)

router = APIRouter(prefix="/sns", tags=["sns"])


# Posts
@router.get("/posts", response_model=PostListResponse)
async def get_posts(
    author_id: Optional[str] = None,
    tag: Optional[str] = None,
    following: bool = False,
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
):
    """Get list of posts with optional filters"""
    raise NotImplementedError


@router.get("/posts/{post_id}", response_model=PostResponse)
async def get_post(post_id: str):
    """Get a specific post"""
    raise NotImplementedError


@router.post("/posts", response_model=PostResponse)
async def create_post(post: PostCreate):
    """Create a new post"""
    raise NotImplementedError


@router.delete("/posts/{post_id}")
async def delete_post(post_id: str):
    """Delete a post"""
    raise NotImplementedError


@router.post("/posts/{post_id}/like")
async def like_post(post_id: str):
    """Like a post"""
    raise NotImplementedError


@router.delete("/posts/{post_id}/like")
async def unlike_post(post_id: str):
    """Unlike a post"""
    raise NotImplementedError


# Comments
@router.get("/posts/{post_id}/comments", response_model=CommentListResponse)
async def get_comments(
    post_id: str,
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
):
    """Get comments for a post"""
    raise NotImplementedError


@router.post("/posts/{post_id}/comments", response_model=CommentResponse)
async def create_comment(post_id: str, comment: CommentCreate):
    """Add a comment to a post"""
    raise NotImplementedError


@router.delete("/posts/{post_id}/comments/{comment_id}")
async def delete_comment(post_id: str, comment_id: str):
    """Delete a comment"""
    raise NotImplementedError


# Follows
@router.post("/users/{user_id}/follow", response_model=FollowResponse)
async def follow_user(user_id: str):
    """Follow a user"""
    raise NotImplementedError


@router.delete("/users/{user_id}/follow")
async def unfollow_user(user_id: str):
    """Unfollow a user"""
    raise NotImplementedError


@router.get("/users/{user_id}/followers", response_model=FollowListResponse)
async def get_followers(
    user_id: str,
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
):
    """Get followers of a user"""
    raise NotImplementedError


@router.get("/users/{user_id}/following", response_model=FollowListResponse)
async def get_following(
    user_id: str,
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
):
    """Get users that a user is following"""
    raise NotImplementedError
